runtime/foucoco/src/xcm_config.rs

```diff
@@ -64,29 +66,31 @@ pub struct CurrencyIdConvert;
 
 impl Convert<CurrencyId, Option<MultiLocation>> for CurrencyIdConvert {
 	fn convert(id: CurrencyId) -> Option<MultiLocation> {
-		match id {
-			CurrencyId::XCM(xcm_assets::RELAY) => Some(MultiLocation::parent()),
-			// Moonbase testnet native token
-			CurrencyId::XCM(xcm_assets::MOONBASE_DEV) => Some(moonbase_alpha::DEV_location()),
-			CurrencyId::Native => Some(native_location_external_pov()),
-			_ => None,
-		}
+		AssetRegistry::metadata(&id)
+			.filter(|m| m.location.is_some())
+			.and_then(|m| m.location)
+			.and_then(|l| l.try_into().ok())
 	}
 }
 
 impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
```

Comment:
We might also want to have a comment similar to [this](https://github.com/centrifuge/centrifuge-chain/blob/47fff268fc7cea828b034a3626df5cc289459a52/runtime/common/src/xcm.rs#L176-L179) here so that it's clear why we are unanchoring the location. 

Actually, I don't think we even need to do this. In Pendulum's asset registry we put 'native' and the Stellar assets as `location: V2: {parents: 0, ... }`. And as far as I understand the code here, this would re-anchor the location from `parents: 0` to `parents: 1`, thus we wouldn't be able to find the asset in the registry anymore. Or am I misunderstanding the logic here? 

We could instead of course change the definitions in the asset registry so that every location is depicted as `parents: 1, ...` but we need to keep this in mind.

Reply:
Yes you got the logic right, and this is a decision we need to make. I assumed (incorrectly) that we were going to add all assets from the relay perspective, that is unanchored. Then as you can also see [here](https://github.com/pendulum-chain/pendulum/pull/432/files#diff-5dc9ab00db16dcf58b342b3f2a9ddd05a114370b9d04b514a3812083382b416dR71) we are converting now the location to currency before fetching the fee. This is mostly because I encountered the problem during testing that the anchored location of `Native` was being used.

But besides tests, I thought it is a good idea to be able to handle both representations. For instance, part of the message may execute from our chain and the representation in this case may be the anchored one.

Since we are bounded by other chains sending assets represented as unanchored, I think that is also the best way to represent our assets (Native and as you mention, Spacewalk). But I don't have too much of a strong opinion. Let me know what you think. Otherwise, we could go the other way around and anchor every asset that is ours.

Reply:
I see. IIRC the assets are re-anchored during the processing of XCM messages, so e.g. assuming that chain B sends us our native token with `MultiLocation { parents: 1, parachain: 2094, palletInstance: xx }`, we'll always see it as `MultiLocation { parents: 0, palletInstance: xx }`. I think this re-anchoring will always happen before our custom logic for handling it is called though we probably want to double-check this. Personally, I think it makes most sense if we have all asset multi-locations represented in their anchored form ie. with `parents: 0` if applicable. I checked other parachains and at least centrifuge does this. The other parachains I checked don't seem to have added their own native tokens to their asset registry. 

We can discuss this in a synchronous meeting or invite the other @pendulum-chain/devs to share their opinions here.

Reply:
I also remember about the re-anchoring, but I am not sure where this happen. I will try to look for it. Also, in centrifuge's they are using the absolute representation and perform the conversion for the `CurrencyIdConvert` type ([here](https://github.com/centrifuge/centrifuge-chain/blob/1a9e385cdc812c4b1a0e55872bbb3576bc48814a/runtime/common/src/xcm.rs#L180))

Reply:
Oh right, thanks for linking to that function again. I forgot about the explanation in the doc comment. This indeed suggests that it arrives unanchored. 

Then I suggest we first try to make clear whether this anchoring happens or not (probably done best by running some of our integration tests with logs) and if we see that no re-anchoring happens we leave it as is and change the locations in our asset registry.

Reply:
Okay so I attempted to use our assets from our chain perspective, yet I hit a roadblock on the following scenario:

Whenever we call `XTokens::transfer` we must specify the asset we want to transfer as `CurrencyId` which the `XTokens` pallet would then [convert](https://github.com/open-web3-stack/open-runtime-module-library/blob/2dbcd1de4db5c05c8596fab447323b4b524673e2/xtokens/src/lib.rs#L409) to a multilocation. Since this now would return a multilocation anchored from our perspective, the execution of such a transfer fails with `Unroutable` error.

This is also probably why previously we [returned](https://github.com/pendulum-chain/pendulum/blob/main/runtime/pendulum/src/xcm_config.rs#L104) the external point of view of our Native token during conversion (thanks @ebma for pointing this out).

So, even though when we receive an asset it is from the native point of view, at least for this scenario described with `XTokens::transfer` we need the asset register to return the external point of view. @TorstenStueber do you think it is okay if we just use the external point of view for the asset registry given this issue?

Reply:
I just read this comments now after I made my review. This explains a little more.

First: let's avoid notions like anchored and unanchored – I don't think they make sense. Every multilocation is relative to something and in the context of our own parachain we should only consider multilocations that are relative to our chain and also use multilocations in the asset registry that are relative to us *without any extra steps one layer up and one layer back down*.

I left some comments to argue that we don't need the `ParachainId` in `CurrencyIdConvert`. Centrifuge might have their own reasons why they do it the way they do it but I don't think that this is a reason for us to copy their behavior, unless we understand their reasoning.

This leaves the problem with `xTokens::transfer`. Can you give a concrete examples that fails? I expect that `xTokens::transfer` works happily with currency ids referenced as (parent: 0, ...) and would assume that we have some misconfiguration here.

Reply:
I had a rough look at how xTokens works and think that the problem occurs [here](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/xtokens/src/lib.rs#L654-L666). Here the XCM message that is executed locally depends on the `transfer_kind`.

Since in this case you intend to move an asset that originates from our chain, that this should behave like a [reserve transfer deposit](https://www.notion.so/satoshipay/23-06-01-Reserve-Transfer-Deposit-910fd65f067c4f2186562494ed7345e8) and should resemble the result of directly calling `pallet-xcm` → `limitedReserveTransferAssets` (I consider xTokens just a set of convenience methods around that).

It becomes clear to me that the behavior we want is the one of the function `transfer_self_reserve_asset` (see https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/xtokens/src/lib.rs#L680-L695) because it is the only one of the three functions that has the same result as `pallet-xcm` → `limitedReserveTransferAssets` (see [here](https://www.notion.so/satoshipay/23-06-01-Reserve-Transfer-Deposit-910fd65f067c4f2186562494ed7345e8?pvs=4#c366724ae9684163b67981655fd7d396)).

We get this behavior when the [call to `transfer_kind`](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/xtokens/src/lib.rs#L649C52-L649C71) returns `SelfReserveAsset`.

However, this only happens if `reserve == self_location` [in this line](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/xtokens/src/lib.rs#L825) where:
* `reserve` is the result of [our configuration `ReserveProvider`](https://github.com/pendulum-chain/pendulum/blob/dd8eb3d55ac2933bfc375a13b7a9d2d73ca4ea86/runtime/pendulum/src/xcm_config.rs#L576) applied to the asset multilocation -> [this will just return (parent: 0, Here)](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/traits/src/location.rs#L68-L69)
* `self_location` is the value of our [`SelfLocation` config](https://github.com/pendulum-chain/pendulum/blob/dd8eb3d55ac2933bfc375a13b7a9d2d73ca4ea86/runtime/pendulum/src/xcm_config.rs#L222) and is obviously different

So we need to ensure that `ReserveProvider` for multilocations of the form (parent: 0, ...) return the same as `SelfLocation`.

Reply:
Looking further, we just need to change the SelfLocation to (parent: 0, Here), just see the last but one paragraph [on this page](https://github.com/open-web3-stack/open-runtime-module-library/blob/4ae0372e2c624e6acc98305564b9d395f70814c0/xtokens/README.md):

> Notice the implementation for now also relies on SelfLocation which is already in xtokens config. The SelfLocation can be set to the absolute view (1, Parachain(THIS_PARACHAIN_ID)) and refers to the sender parachain. The alternative is to set SelfLocation to relative view (0, Here) to adhere to Polkadot guidelines.

Reply:
Thanks for the investigation, that certainly solves the issue faced by `xTokens::transfer()`, representing our native as `(parent: 0, Here)` introduces some other breaking changes for our current test setup. I will try to explore what exactly fails now. Might have to do with our sibling chain, since now the asset it receives is different.

Reply:
Can you elaborate? Maybe I can help.

Reply:
So here I pushed a last [commit](https://github.com/pendulum-chain/pendulum/pull/432/commits/ff45e0c5c552efe8671328e5c0bd5e611af9834e) that implements this idea. I tested both `transfer` and `transfer_multiassets` and they are both working. There were some extra changes to be made though:

- First, in the sibling chain when sending (parents: 0, Here) we now receive the following [multilocation](https://github.com/pendulum-chain/pendulum/pull/432/commits/ff45e0c5c552efe8671328e5c0bd5e611af9834e#diff-84e3c62ba95e693cd5e55dd32c931b4d8ebad427f21b49e2442d288205e24c11R154), which we must handle. I hope this does not break other integrations that other parachains have with us.

- Secondly, and this is only regarding our chain, I had to update a bit the [convert](https://github.com/pendulum-chain/pendulum/pull/432/commits/ff45e0c5c552efe8671328e5c0bd5e611af9834e#diff-3287a4ee060b1070a8e6559a094b4890a596a1b281effef54cafe59da20edf49R247) function. Since `transfer_multiassets` may send already the location from an external point of view, we also need to handle this conversion, that is why we anchor (yes I know the nomenclature is loose but is the name of the method) to our chain all assets before converting. 

- Finally on the same function, we handle the case where the native token recieved is represented as `(0,X1(PalletInstance(10)`, this we cannot control since it comes from other parachains. In this case the testing sibling.

All the changes to the `convert` function are needed since we are now charging the fee on the incoming asset we need to first convert it to access our `asset_metadata`. 

Aside from point 1 which I am a bit afraid may break how other parachains receive our Native, this could work.

Reply:
I'm not sure I follow you. The tests were just wrong in my opinion, it does not correct to – in the context of our parachain – ever refer to an asset as (parent: 1, parachain(id), x) if "id" is our own parachain, instead we need to refer to it as (parent: 0, x).

I don't understand why you would do any reanchoring in `CurrencyIdConvert`. In what context is that required?

Reply:
I agree that the test were wrong, in a way. I think it is a matter of choice how to refer to our assets, but now it makes more sense to do it from our perspective. Before, it was consistent with our definition of `SelfLocation` like you were able to find. Now, they agree with the local view.

With respect to the reanchoring, as I said I don't know of anything that would need it right now that is why we can remove it. But since it happened before with pallet `xTokens` which was trying to convert assets represented from an external point of view I imagined this could happen by some other pallet that also uses the convert function.

Reply:
Are there other pallets that use the convert function?

Reply:
I will look at that tomorrow and if there are not, then let's remove it!

Reply:
Nice, thanks.